### PR TITLE
8254024: Enhance native libs for AWT and Swing to work with GraalVM Native Image

### PR DESCRIPTION
**Pull request by arodionov**
The following PR fixes https://bugs.openjdk.java.net/browse/JDK-8254024

Starting from version 11.0.9, all JDK libraries also build as static libraries (JEP 178: Statically-Linked JNI Libraries).
The purpose of using static libraries is to build GraalVM Native image statically linked with Java native libraries, to shipping single executable without runtime dependencies from JRE.

For some static libraries, it leads to an issue: if one static library is trying to load another static library in runtime using `dlopen` (and then uses `dlsym` to find symbols). With static libraries, this is not possible and all dynamic calls should be replaced.

This happens in `libawt` while it loads `awt-xawt` or `awt_headless` and in `mlib_image`.

Current PR fixes this issue for AWT libraries that allow building Swing/AWT applications as a Native image and have no runtime dependencies from JRE.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8254024](https://bugs.openjdk.java.net/browse/JDK-8254024): Enhance native libs for AWT and Swing to work with GraalVM Native Image

### Reviewers
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)
 * [Bob Vandette](https://openjdk.java.net/census#bobv) (@bobvandette - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/562/head:pull/562`
`$ git checkout pull/562`

### CODE REVIEW

**Comment by robilad** [comment]
Hi, please send the details whose OCA you are covered under to  Dalibor.topic@oracle.com so that I can verify your account. Thanks!

**Comment by arodionov** [comment]
@robilad Hi Dalibor! Have sent you email with details.
Thanks!

**Comment by arodionov** [comment]
@bobvandette Could you please join to the review?

**Comment by prrace** [comment]
It is generally not expected for such a proposed change to pop up out of nowhere from someone who is a new contributor
without prior discussion and so forth.
I have no idea where this change is coming from, why we would want it,  and how it is expected to be supported and you have provided zero
explanation of any of it in the PR. I think you need to start by explaining the big picture, and what you expect to do
with this and how you are guaranteeing to support it and ensure it breaks NOTHING else. At all.
And then we could also use some technical details on what you perceive is needed and why this resolives it.
Also is this "everything" ?

File: src/java.desktop/unix/native/libawt_xawt/xawt/XToolkit.c
Comment (bobvandette) [inline src/java.desktop/unix/native/libawt_xawt/xawt/XToolkit.c:820]: I'm not sure if this entrypoint is ever accessed externally in a normal JDK.  To be safe, can you make both name changes conditional on STATIC_BUILD?

**Comment by bobvandette** [comment]
> It is generally not expected for such a proposed change to pop up out of nowhere from someone who is a new contributor
> without prior discussion and so forth.
> I have no idea where this change is coming from, why we would want it, and how it is expected to be supported and you have provided zero
> explanation of any of it in the PR. I think you need to start by explaining the big picture, and what you expect to do
> with this and how you are guaranteeing to support it and ensure it breaks NOTHING else. At all.
> And then we could also use some technical details on what you perceive is needed and why this resolives it.
> Also is this "everything" ?

I've been coaching @arodionov through this change.  This fix extends the work I did to allow a Java program to be created using all static libraries.

**Comment by prrace** [comment]
Sorry I am not familiar with that work. Please provide more info.

**Comment by bobvandette** [comment]
> Sorry I am not familiar with that work. Please provide more info.

Here's the JEP for the work done in JDK8 to allow libraries to be built and used statically.

https://bugs.openjdk.java.net/browse/JDK-8046168

Here's the changes done to allow a Mac build to be done fully statically.

http://hg.openjdk.java.net/jdk9/jdk9/jdk/rev/55573c377d64

File: src/java.desktop/unix/native/libawt_xawt/xawt/XToolkit.c
Comment (arodionov) [reply src/java.desktop/unix/native/libawt_xawt/xawt/XToolkit.c:821]: Have made both names conditional.

**Comment by mrserb** [comment]
> 
> Here's the JEP for the work done in JDK8 to allow libraries to be built and used statically.

Does it mean that in the case of static-libs both variants of AWT libraries(xawt and headless) will be loaded?
How it is possible to test the current changes?

**Comment by prrace** [comment]
I would have to suppose that isn't possible but it is not clear whether the authors
of the fix are OK with this or not because the PR is very light on background and explanation. Which is a blocker for me.

I don't get the motivation from the JEP as to how it applies here.
There are two motivations listed
1) For environments that don't support dynamic linking, and that doesn't apply to 64 bit Linux server so we can rule that out
2) For applications that "want" to statically link. I want a lot of things but wanting it isn't a very well expressed motivation for something like this.

**Comment by arodionov** [comment]
> Does it mean that in the case of static-libs both variants of AWT libraries(xawt and headless) will be loaded?
No, they shouldn't. If this happens it causes duplicate symbol error during linkage phase.
For GraalVM native image we select what library should be linked based on `-Djava.awt.headless` property.

>  How it is possible to test the current changes?
For GraalVM native image, we run JCK tests `api/java_awt` and `api/javax_swing`. Those tests built as a native image.

**Comment by mrserb** [comment]
> > How it is possible to test the current changes?
> 
> For GraalVM native image, we run JCK tests `api/java_awt` and `api/javax_swing`. Those tests built as a native image.

I meant some "steps to reproduce" the current bug, and which can be used to verify the current fix.

**Comment by arodionov** [comment]
> > > How it is possible to test the current changes?
> > 
> > 
> > For GraalVM native image, we run JCK tests `api/java_awt` and `api/javax_swing`. Those tests built as a native image.
> 
> I meant some "steps to reproduce" the current bug, and which can be used to verify the current fix.

Actually, we have found this issue during JCK tests execution that was built as a native image. The fix criteria are the successful tests run.

**Comment by arodionov** [comment]
Yes, I understand what you mean.
Unfortunately, I don't know a straightforward way.
Maybe the easiest way to reproduce an error and validate the fix is to download the future version of GraalVM with AWT/Swing support (hope it will be at the end of October), replace `libawt.a` to its unpatched library version and then build a native image based on the following code:
```
        float[] data = new float[]{0.0f, 1.1f, 2.2f, 3.3f, 4.4f, 5.5f, 6.6f, 7.7f, 8.8f};
        Kernel kernel = new Kernel(3, 3, data);
        ConvolveOp op = new ConvolveOp(kernel);

        BufferedImage src = new BufferedImage(10, 10, BufferedImage.TYPE_INT_ARGB);
        BufferedImage dest = new BufferedImage(10, 10, BufferedImage.TYPE_INT_ARGB);

        op.filter(src, dest);
```
The built image run should throw "Unable to convolve src image" ImagingOpException exception.

**Comment by arodionov** [comment]
In general, the 2 following points should be changed and tested in the static libraries:
- rename conflicting/duplicate symbols in libraries that are linked together
- have a workaround for the functionality implemented via dlopen/dlsym

**Comment by arodionov** [comment]
@mrserb With the last dev build of GraalVM based on JDK 11.09, the bug can be reproduced by the approach that I've mentioned above.
The general steps: 
- download the [last dev build](https://github.com/graalvm/graalvm-ce-dev-builds/releases) of GraalVM + Native image 
- build the native image for AWTFixExample class (ziped configs for reflection and JNI is attached): 
`$ ~/graalvm-ce-java11-21.0.0-dev/bin/native-image -H:ConfigurationFileDirectories=./configs -Djava.awt.headless=true --native-image-info AWTFixExample`
- run built image - it should be executed successfully
- clone any version of the OpenJDK (11.09 - 16)
- build static libraries by executing `make static-libs-image` target 
- replace `libawt.a` in GraalVM on built static library from OpenJDK
- try to build example again and run it - it should be failed
- clone OpenJDK with the proposed fix
- build static libraries
- replace `libawt.a` in GraalVM on built static library with the fix
- build example and run it - it should be executed successfully

```
public class AWTFixExample {
    public static void main(String[] args) {
        float[] data = new float[]{0.0f, 1.1f, 2.2f, 3.3f, 4.4f, 5.5f, 6.6f, 7.7f, 8.8f};
        Kernel kernel = new Kernel(3, 3, data);
        ConvolveOp op = new ConvolveOp(kernel);

        BufferedImage src = new BufferedImage(10, 10, BufferedImage.TYPE_INT_ARGB);
        BufferedImage dest = new BufferedImage(10, 10, BufferedImage.TYPE_INT_ARGB);

        BufferedImage image = op.filter(src, dest);
        System.out.println(image);
    }
}
```

[configs.zip](https://github.com/openjdk/jdk/files/5490906/configs.zip)

**Comment by mrserb** [comment]
I think part of the changed code is actually dead code and can be removed, I'll take a look.

**Comment by mrserb** [comment]
The changes in awt_LoadLibrary.c and XToolkit.c is mostly noop except excluding dladdr/dlopen. Since XsessionWMcommand/AWTIsHeadless are not used at least in the mainline(jdk16).
The changes in the mlib look fine.

The testing is in progress.

**Comment by mrserb** [comment]
please merge the master to this branch, it is quite outdated, and update the last copyright date to 2020.

**Comment by arodionov** [comment]
@mrserb I have rebased PR on the master, but `XsessionWMcommand` and `AWTIsHeadless` functions are still at `XToolkit.c` in the mainline. So, to prevent the conflicting symbols error during static libraries linkage (`libawt.a` and `libawt_xawt.a`), we need to rename or to hide conflicting functions.

**Review by mrserb — APPROVED** [review]

**Comment by arodionov** [comment]
@dougxc could you please review AWT static-libs enhancement for GraalVM Native Image?

**Comment by dougxc** [comment]
> @dougxc could you please review AWT static-libs enhancement for GraalVM Native Image?

Sorry, this needs to be reviewed by people who understand this area of the code. Maybe @mrserb can suggest someone.

**Comment by arodionov** [comment]
@prrace I have added more information to PR description, sorry for the inconvenience. 

Regarding the use case of using AWT static libraries: with GraalVM Native image we can take a jar with some AWT/Swing application and convert it (using gcc and static libraries) to binary executable (currently only for Linux 64) and then run it without JRE.

**Comment by prrace** [comment]
> This happens in libawt while it loads awt-xawt or awt_headless and in mlib_image.

So I am supposing this means you are producing two different static images ? - either all the time or based on a build flag ?
One is linking in the X11 lib, the other the headless stub lib ??

Can you confirm that understanding ?
This of course doesn't scale very well - meaning you are lucky we don't do this more often  as you may need more combinations.

**Comment by arodionov** [comment]
No, we produce only one image. The headless/non-headless mode is defined by the system property `-Djava.awt.headless=...`.
Here is a command example for native image generation:
`$ ~/graalvm-ce-java11-21.0.0-dev/bin/native-image -H:ConfigurationFileDirectories=./configs -Djava.awt.headless=true AWTFixExample`

With `-Djava.awt.headless=false` to the gcc linker we pass `libawt.a` and `libawt_headless.a`
With `-Djava.awt.headless=true` to the gcc linker we pass `libawt.a` and `libawt-xawt.a`
If the generated image demands `libmlib_image.a` it will be also linked.

**Comment by prrace** [comment]
> No, we produce only one image. The headless/non-headless mode is defined by the system property `-Djava.awt.headless=...`.
> Here is a command example for native image generation:
> `$ ~/graalvm-ce-java11-21.0.0-dev/bin/native-image -H:ConfigurationFileDirectories=./configs -Djava.awt.headless=true AWTFixExample`
> 
> With `-Djava.awt.headless=false` to the gcc linker we pass `libawt.a` and `libawt_headless.a`
> With `-Djava.awt.headless=true` to the gcc linker we pass `libawt.a` and `libawt-xawt.a`
> If the generated image demands `libmlib_image.a` it will be also linked.

-Djava.awt.headless is a runtime property not a build time property. So you are doing domething very weird.
How does that image with "false" run a UI app ? It can't. And yet the one with "true" still can run without an xserver, except it will not start unless you have X11 libs installed - and I am *reasonably* sure you aren't statically linking all the system provided libraries ... I hope ...

So how does a "user" use all of this ? Seems they must create their own app by native compile + linking.
Can they create an image that can run any java app, or must the app like AWTFixExample be linked in to the image ?

**Comment by arodionov** [comment]
At the current stage, we made the `-Djava.awt.headless` option be used during the build time, at the same time it does no influence at the runtime. The user should know the environment where it is planned to run a native image in advance. It is his responsibility to pass correct options to the native image compiler.

For example, for the code snippet `AWTFixExample` (mentioned above https://github.com/openjdk/jdk/pull/562#issuecomment-722023476), that is headless, the following set of static libraries are used for image build:
```
# Static libraries:
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/svm/clibraries/linux-amd64/liblibchelper.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/liblcms.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libawt_headless.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libawt.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libmlib_image.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libnet.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libnio.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libjava.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libfdlibm.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libzip.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/svm/clibraries/linux-amd64/libjvm.a
# Other libraries: stdc++,m,pthread,dl,z,rt
```
For simple AWT UI Java app, the set of libraries are the following:
```
# Static libraries:
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libnet.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libjavajpeg.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libnio.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/svm/clibraries/linux-amd64/liblibchelper.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libjava.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/liblcms.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libfontmanager.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libawt_xawt.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libawt.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libfdlibm.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/static/linux-amd64/glibc/libzip.a
#   ../../vms/graalvm-ce-java11-21.0.0-dev/lib/svm/clibraries/linux-amd64/libjvm.a
# Other libraries: X11,Xrender,Xext,Xi,stdc++,m,freetype,pthread,dl,z,rt
```

> So how does a "user" use all of this ? Seems they must create their own app by native compile + linking.
> Can they create an image that can run any java app, or must the app like AWTFixExample be linked in to the image ?

The global idea is that any existing Java app (even without source code) can be converted into a native image, that's why we have added AWT/Swing support to GraalVM.
To make a native image from the existing Java app, the user needs to have GraalVM, gcc with the set of libraries (platform dependent).
After running `native-image` command (from GraalVM), with a set of options and additional configuration files, and specifying the target jar or class-file, an executable binary for the current platform will be built.
There are limitations and the list of unsupported features (https://www.graalvm.org/reference-manual/native-image/Limitations/) for Java apps, so it may demand re-writing some code or providing so-called substitutions to make it compatible for the native-image compiler.

**Comment by arodionov** [comment]
@prrace It may be also useful to look at how we define a set of static libraries:
https://github.com/oracle/graal/blob/9390324021ab7b6ed391f2c8f62be490a7ab13db/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationAwt.java#L104

**Review by magicus — APPROVED** [review]
This makes sense from a build PoV, but you'll need approval from the client team as well.

**Comment by prrace** [comment]
> @prrace It may be also useful to look at how we define a set of static libraries:
> https://github.com/oracle/graal/blob/9390324021ab7b6ed391f2c8f62be490a7ab13db/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationAwt.java#L104

If by "help" you mean scare the life out of me ..
I am sure I don't understand all of it but enough to know  I've never seen so many references to JDK internals and behaviours
by some code that is not part of the JDK itself.
It seems very fragile to the point of not being supportable.

If someone else wants to approve this change they can, but my conscience isn't letting me ..

**Review by bobvandette — APPROVED** [review]
I'm ok with these changes.  They are no worse than the other STATIC_BUILD changes that already exist in the JDK sources.

**Comment by vjovanov** [comment]
@prrace I agree with you that our intrusion is a large maintenance overhead and that it can lead to potential errors. It is worth mentioning two things that could make you feel less scared: 
1) Much of this is code temporary: the code is here until we find a proper way to integrate this information into the JDK itself, without incurring a noticeable maintenance overhead for the JDK maintainers.
2) We test all our code intensively with all the tests provided by the JDK assuring that on every release we pass the same tests that the JDK is passing.

**Pull request closed without merging**
